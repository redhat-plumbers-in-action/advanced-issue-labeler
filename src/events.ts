//? DOC: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads

export const events = {
  issue: ['issues.opened' as const],
};
