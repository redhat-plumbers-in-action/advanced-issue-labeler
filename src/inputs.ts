import { IsArray, IsString, MinLength, ValidateIf } from 'class-validator';

import { TInputs } from './types.d';

export class Inputs {
  @ValidateIf(instance => instance._template)
  @IsString()
  @MinLength(1)
  private _template?: string | undefined;

  @IsString()
  @MinLength(1)
  private _section?: string | undefined;

  @ValidateIf(instance => instance._blockList)
  @IsArray()
  @IsString({ each: true })
  @MinLength(1, { each: true })
  private _blockList?: string[] | undefined;

  constructor(inputs: TInputs) {
    this._template = inputs?.template;
    this._section = inputs?.section;
    this._blockList = inputs?.blockList;
  }

  get template() {
    return this._template;
  }

  get section() {
    return this._section;
  }

  get blockList() {
    return this._blockList;
  }
}
