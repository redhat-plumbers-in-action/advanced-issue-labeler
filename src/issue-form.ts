export class IssueForm {
  private _parsed: { [key: string]: string };

  constructor(parsedIssue: {}) {
    this._parsed = parsedIssue;
  }

  private get parsed() {
    return this._parsed;
  }

  isProperty(key: string) {
    return this.parsed.hasOwnProperty(key);
  }

  getProperty(key: string) {
    return this.parsed[key];
  }

  getSafeProperty(key: string) {
    if (!this.isProperty(key)) return undefined;

    return this.getProperty(key);
  }

  listKeywords(key: string, blockList: string[]) {
    const propertySection = this.getSafeProperty(key);

    if (!propertySection) return undefined;

    return propertySection
      .split(', ', 25)
      .filter(label => !this.isCompliant(blockList, label));
  }

  isCompliant(policy: string[], keyword: string) {
    return !!policy.find(rule => keyword === rule);
  }
}
