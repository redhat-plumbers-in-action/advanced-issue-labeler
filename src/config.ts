import { debug } from '@actions/core';
import { Context } from 'probot';
// import { Type } from 'class-transformer';
import {
  ArrayMinSize,
  IsArray,
  IsString,
  MinLength,
  validate,
  ValidateNested,
} from 'class-validator';

import { events } from './events';
import {
  TConfigObject,
  TPolicyItem,
  TSectionItem,
  TLabelItem,
} from './types.d';
import { ValidationFeedback } from './validation-feedback';

export class Config {
  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  // @Type(() => SectionItem)
  private _policy: PolicyItem[];

  constructor(config: TConfigObject) {
    this._policy = Array.isArray(config?.policy)
      ? config.policy.map(item => new PolicyItem(item))
      : [];
  }

  get policy() {
    return this._policy;
  }

  getTemplatePolicy(template: string | undefined) {
    // ?: When template name is provided look for section that match the template name
    const templatePolicy = this.policy.find(pItem =>
      pItem.template?.find(tItem => tItem === template)
    );

    // If exact policy for template exists ; return it
    if (templatePolicy) return templatePolicy;

    // ?: When template name is undefined look for section that undefined template field
    debug(`Looking for default policy ...`);
    return this.policy.find(
      pItem =>
        !pItem?.template ||
        (Array.isArray(pItem?.template) && pItem?.template.length === 0)
    );
  }

  static async getConfig(
    context: {
      [K in keyof typeof events]: Context<typeof events[K][number]>;
    }[keyof typeof events]
  ) {
    const retrievedConfig = await context.config<TConfigObject>(
      'advanced-issue-labeler.yml'
    );

    if (Config.isConfigEmpty(retrievedConfig)) {
      return null;
    }

    const config = new this(retrievedConfig as Config);

    return config;
  }

  static isConfigEmpty(config: TConfigObject | null | unknown) {
    return config === null;
  }

  static async validate(instance: Config) {
    const validationResult = await validate(instance, {
      whitelist: true,
      forbidNonWhitelisted: true,
    });

    const results = validationResult.map(error => {
      return ValidationFeedback.composeFeedbackObject(error);
    });

    return results;
  }
}

class PolicyItem {
  @IsString({ each: true })
  @MinLength(0, { each: true })
  private _template?: string[];

  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  // @Type(() => SectionItem)
  private _section: SectionItem[];

  constructor(item: TPolicyItem) {
    this._template = item?.template ?? [];
    this._section = Array.isArray(item?.section)
      ? item.section.map(sectionItem => new SectionItem(sectionItem))
      : [];
  }

  get template() {
    return this._template;
  }

  get section() {
    return this._section;
  }
}

class SectionItem {
  @IsString({ each: true })
  @MinLength(1, { each: true })
  private _id: string[];

  @IsString({ each: true })
  @MinLength(0, { each: true })
  private _blockList: string[];

  @IsArray()
  @ValidateNested({ each: true })
  @ArrayMinSize(1)
  // @Type(() => SectionItem)
  private _label: TLabelItem[];

  constructor(item: TSectionItem) {
    this._id = item?.id;
    this._blockList = item['block-list'] ?? [];
    this._label = Array.isArray(item?.label)
      ? item.label.map(labelItem => new Label(labelItem))
      : [];
  }

  get id() {
    return this._id;
  }

  get blockList() {
    return this._blockList;
  }

  get label() {
    return this._label;
  }
}

class Label {
  @IsString()
  @MinLength(1)
  private _name: string;

  @IsString({ each: true })
  @MinLength(1, { each: true })
  private _keys: string[];

  constructor(item: TLabelItem) {
    this._name = item?.name;
    this._keys = item?.keys;
  }

  get name() {
    return this._name;
  }

  get keys() {
    return this._keys;
  }
}
