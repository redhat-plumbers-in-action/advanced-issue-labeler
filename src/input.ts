import { z } from 'zod';
import { IssueForm } from './issue-form';
import { inputSchema, InputType, jsonSchema } from './schema/input';

export class Input {
  readonly issueForm: InputType['issueForm'];
  readonly template: InputType['template'];
  readonly section: InputType['section'];
  readonly blockList: InputType['blockList'];

  constructor(input: unknown) {
    const parsedInput = inputSchema.parse(input);

    this.issueForm = parsedInput.issueForm;
    this.template = parsedInput.template;
    this.section = parsedInput.section;
    this.blockList = parsedInput.blockList;
  }
}
