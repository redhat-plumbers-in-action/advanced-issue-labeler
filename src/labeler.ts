import { debug } from '@actions/core';
import { ValidateIf, validate, ValidateNested, Allow } from 'class-validator';

import { Config } from './config';
import { IssueForm } from './issue-form';
import { ValidationFeedback } from './validation-feedback';
import { TInputs } from './types.d';

export class Labeler {
  @ValidateIf(instance => !instance._inputs)
  @ValidateNested()
  private _config?: Config | undefined | null;

  @ValidateIf(instance => !instance._config)
  @ValidateNested()
  private _inputs?: TInputs;

  @Allow()
  private _issueForm: IssueForm;

  // State holders
  @Allow()
  private _isConfig?: boolean = undefined;
  @Allow()
  private _isInputs?: boolean = undefined;

  constructor(issueForm: IssueForm) {
    this._issueForm = issueForm;
  }

  get config() {
    return this._config;
  }

  set config(config: Config | undefined | null) {
    this._config = config;
  }

  get inputs() {
    return this._inputs;
  }

  set inputs(inputs: TInputs | undefined) {
    this._inputs = { ...inputs };
  }

  get isConfig() {
    if (this._isConfig === undefined) this.isConfig = !!this.config;

    return this._isConfig;
  }

  set isConfig(value: boolean | undefined) {
    this._isConfig = value;
  }

  get isInputs() {
    if (this._isInputs === undefined) {
      // TODO: This needs rework!
      // template is required
      this.isInputs = !!this.inputs?.section && this.inputs.section !== '';
    }

    return this._isInputs;
  }

  set isInputs(value: boolean | undefined) {
    this._isInputs = value;
  }

  private get issueForm() {
    return this._issueForm;
  }

  private set issueForm(issueForm: IssueForm) {
    this._issueForm = issueForm;
  }

  gatherLabels() {
    if (this.isInputs) {
      return this.inputBasedLabels();
    }

    return this.configBasedLabels();
  }

  // Expects that validation method was called in advance
  private inputBasedLabels() {
    const keywords = this.issueForm.listKeywords(
      this.inputs?.section ?? '',
      this.inputs?.blockList ?? ['']
    );

    if (!keywords || keywords?.length === 0) {
      debug(`Section field is empty.`);
      return;
    }

    if (!keywords[0]) {
      debug(`Section field is empty.`);
      return;
    }

    return keywords;
  }

  private configBasedLabels() {
    const labels: string[] = [];

    // Pick correct policy from config based on template input
    const selectedPolicy = this.config?.getTemplatePolicy(
      this.inputs?.template
    );

    if (!selectedPolicy) {
      debug(`Policy wasn't provided!`);
      return;
    }

    // For each section of policy
    for (const sectionItem of selectedPolicy.section) {
      // For each id of policy
      for (const singleID of sectionItem.id) {
        if (!this.issueForm.isProperty(singleID)) {
          debug(`Issue form doesn't contain section: ${singleID}`);
          continue;
        }

        const keywords = this.issueForm.listKeywords(
          singleID,
          sectionItem.blockList
        );

        if (!keywords || keywords?.length === 0) {
          debug(`Section field is empty.`);
          continue;
        }

        if (!keywords[0]) {
          debug(`Section field is empty.`);
          continue;
        }

        // Apply policy on provided data via issue
        for (const rule of sectionItem.label) {
          if (
            keywords.find(keyword => rule.keys.find(key => keyword === key))
          ) {
            labels.push(rule.name);
          }
        }
      }
    }

    return labels;
  }

  static async validate(instance: Labeler) {
    const validationResult = await validate(instance, {
      whitelist: true,
      forbidNonWhitelisted: true,
    });

    const results = validationResult.map(error => {
      return ValidationFeedback.composeFeedbackObject(error);
    });

    return results;
  }
}
