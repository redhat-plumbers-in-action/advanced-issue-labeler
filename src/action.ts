import { getInput, debug, info } from '@actions/core';

import { Config } from './config';
import { IssueForm } from './issue-form';
import { Labeler } from './labeler';
import { CustomOctokit } from './octokit';
import { context } from '@actions/github';

async function action(octokit: CustomOctokit) {
  const issueFormInput = getInput('issue-form', { required: true });

  const labeler = new Labeler(new IssueForm(JSON.parse(issueFormInput)));
  labeler.config = await Config.getConfig(octokit);

  // If no config was provided try inputs
  if (!labeler.isConfig) {
    labeler.setInputs({
      section: getInput('section', { required: true }),
      blockList: getInput('block-list').split('\n', 25),
    });
  }

  // Config requires template as well
  labeler.setInputs({
    template: getInput('template'),
  });

  const validationResult = await Labeler.validate(labeler);

  if (validationResult.length > 0) {
    for (const [index, message] of validationResult.entries()) {
      // TODO: Probably needs some care:
      debug(
        `{ property: ${message.property}, value: ${message.value}, notes: ${message.notes}`
      );
    }
    return;
  }

  const labels = labeler.gatherLabels();

  // Check if there are some labels to be set
  if (!labels || (Array.isArray(labels) && labels?.length < 1)) {
    info('Nothing to do here. CY@');
    return;
  }

  info(`Labels to be set: ${labels}`);

  const response = await octokit.request(
    'POST /repos/{owner}/{repo}/issues/{issue_number}/labels',
    {
      ...context.repo,
      issue_number: context.issue.number,
    }
  );

  debug(`GitHub API response status: [${response.status}]`);
}

export default action;
