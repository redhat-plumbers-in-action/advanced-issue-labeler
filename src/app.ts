import { Context, Probot } from 'probot';
import { getInput, debug } from '@actions/core';

import { events } from './events';
import { Config } from './config';
import { IssueForm } from './issue-form';
import { Labeler } from './labeler';

const app = (probot: Probot) => {
  probot.on(
    events.issue,
    async (context: Context<typeof events.issue[number]>) => {
      const issueFormInput = getInput('issue-form');

      if (!issueFormInput) {
        context.log.error(`Parameter issue-form is required!`);
        return;
      }

      const labeler = new Labeler(new IssueForm(JSON.parse(issueFormInput)));
      labeler.config = await Config.getConfig(context);

      // If no config was provided try inputs
      if (!labeler.isConfig) {
        labeler.inputs = {
          section: getInput('section'),
          blockList: getInput('block-list').split('\n', 25),
        };
      }

      // Config requires template as well
      labeler.inputs = {
        template: getInput('template'),
      };

      const validationResult = await Labeler.validate(labeler);

      if (validationResult.length > 0) {
        for (const [index, message] of validationResult.entries()) {
          // TODO: Probably needs some care:
          probot.log.info(
            `{ property: ${message.property}, value: ${message.value}, notes: ${message.notes}`
          );
          probot.log.warning(`[${index}] ~ Config validation: '${message}'`);
        }
        return;
      }

      const labels = labeler.gatherLabels();

      // Check if there are some labels to be set
      if (!labels || (Array.isArray(labels) && labels?.length < 1)) {
        debug('Nothing to do here. CY@');
        return;
      }

      debug(`Labels to be set: ${labels}`);

      const response = await context.octokit.rest.issues.addLabels(
        context.issue({ labels })
      );

      debug(`GitHub API response: ${response}`);
    }
  );
};

export default app;
