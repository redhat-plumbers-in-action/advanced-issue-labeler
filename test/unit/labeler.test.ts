import { describe, it, expect, beforeEach } from 'vitest';

import { IssueForm } from '../../src/issue-form';
import { Labeler } from '../../src/labeler';

describe('IssueForm Object', () => {
  let labelers: Labeler[];

  beforeEach(() => {
    labelers = [new Labeler(new IssueForm({}))];
  });

  it('can be instantiated', () => {
    labelers.map(item => expect(item).toBeDefined());
  });
});
