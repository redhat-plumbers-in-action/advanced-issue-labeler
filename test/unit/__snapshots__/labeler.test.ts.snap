// Vitest Snapshot v1

exports[`IssueForm Object > gatherLabels() 1`] = `undefined`;

exports[`IssueForm Object > get config() 1`] = `undefined`;

exports[`IssueForm Object > get inputs() 1`] = `
Inputs {
  "_blockList": undefined,
  "_section": undefined,
  "_template": undefined,
}
`;

exports[`IssueForm Object > get isConfig() 1`] = `false`;

exports[`IssueForm Object > get isInputs() 1`] = `false`;

exports[`IssueForm Object > is invalid 1`] = `
[
  {
    "notes": {
      "isString": "_section must be a string",
      "minLength": "_section must be longer than or equal to 1 characters",
    },
    "property": "_inputs._section",
    "value": undefined,
  },
]
`;

exports[`IssueForm Object > is invalid 2`] = `
[
  {
    "notes": {
      "isString": "_section must be a string",
      "minLength": "_section must be longer than or equal to 1 characters",
    },
    "property": "_inputs._section",
    "value": undefined,
  },
]
`;

exports[`IssueForm Object > is invalid 3`] = `
[
  {
    "notes": {
      "isString": "_section must be a string",
      "minLength": "_section must be longer than or equal to 1 characters",
    },
    "property": "_inputs._section",
    "value": undefined,
  },
]
`;

exports[`IssueForm Object > is invalid 4`] = `
[
  {
    "notes": {
      "isString": "_section must be a string",
      "minLength": "_section must be longer than or equal to 1 characters",
    },
    "property": "_inputs._section",
    "value": undefined,
  },
]
`;
